linkerd/http-route/src/http/match/path.rs

```diff
@@ -0,0 +1,153 @@
+use http::Uri;
+use regex::Regex;
+
+#[derive(Clone, Debug)]
+pub enum MatchPath {
+    Exact(String),
+    Prefix(String),
+    Regex(Regex),
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub(crate) enum PathMatch {
+    Exact(usize),
+    Regex(usize),
+    Prefix(usize),
+}
+
+// === impl MatchPath ===
+
+impl MatchPath {
+    pub(crate) fn match_length(&self, uri: &Uri) -> Option<PathMatch> {
+        match self {
+            Self::Exact(s) => {
+                if s == uri.path() {
+                    return Some(PathMatch::Exact(s.len()));
+                }
+            }
+
+            Self::Regex(re) => {
+                if let Some(m) = re.find(uri.path()) {
+                    let len = uri.path().len();
+                    // Check that the regex is anchored at the start and end of
+                    // the value.
+                    if m.start() == 0 && m.end() == len {
+                        return Some(PathMatch::Regex(len));
+                    }
```

Comment:
nit: could we change this so that we always construct anchored regexen, and then just use `Regex::is_match` instead? the `regex` crate's docs suggest that `is_match` may be more efficient than `find`.

Reply:
> could we change this so that we always construct anchored regexen

While I think I could get something that would anchor some regexes, I'm not confident that my solution wouldn't have some unexpected corner cases. This solution is at least simple and relatively obvious--it seems unlikely there could be false positives,at least. We can always improve this later if we find it's worth the effort.